Simple generation

davidlgj opened the pull request:
My first stab at a prototype that generates a field from a view shcema and validates it with a validation schema.
It only supports inputs and texareas and lacks a lot at the moment, like not showing error messages when it's pristine, but it's a start. 

It has a starting point directive called forms-js directive that is the startpoint. It kicks of by instantiating a `formsjs.Form` and then builds up child directives depending on `inputType` in the view schema.

``` html
<form forms-js view="view" validation="validation" model="model"></form>
```

The child directives, `fjsText` and `fjsTextarea` then are responsible for building it's part of the form. They both use `ngModel` internally and a `fjsValidate` directive that registers the field with the `formsjs.Form` instance we have.

See `examples/index.html` for more.

davidlgj committed:
Started with a simple form generation

Started in ES5 js and not typescript just to get started. It's rather naive
implementation that renders a form by having a directive for each inputType
in the view schema. Just a standard text input right now.

davidlgj committed:
Basic validation

Still very rough though.

davidlgj committed:
Started the transformation to typescript

davidlgj committed:
Hooked up a formjs.Form and registered attributes.

Still haven't done a submit flow.

davidlgj committed:
Added a textarea as well.

bvaughn commented on source/directives/fjs.ts:
```diff
@@ -0,0 +1,37 @@
+          var view  = values[0];
+          var validation = values[1];
+          if (view !== undefined && validation !== undefined) {
+
```
I assume we're only _requiring_ a view/layout schema in this spike (and it will become optional down the road).

davidlgj replied:
Yes exactly.

I'd like to support three scenarios:
1. Generate it all <- this is what I've done so far
2. Don't give a view definition but give a validation schema

``` html
<form forms-js validation="validation">
<fieldset>
  <legend>Reqnuired fields</legend>
  <fjs-text key="name"></fjs-text>  
  <fjs-text key="email"></fjs-text>
</fieldset>
<fjs-textarea key="comment"></fjs-textarea>
</form>
```
1. Neither, put validation on each field (Not sure about this.. but hey maybe) I guess we would build a validation schema internally.

``` html
<form forms-js>
<fieldset>
  <legend>Reqnuired fields</legend>
  <fjs-text min="3" required="true" key="name"></fjs-text>  
  <fjs-text pattern=".+@.+\..+" key="email"></fjs-text>
</fieldset>
<fjs-textarea max="270" key="comment"></fjs-textarea>
</form>
```

bvaughn replied:
Interesting. I had not considered use-case 3 but maybe there's value there? Seems like it should be doable if we decide there is.

davidlgj committed:
Use AttributeMetadata.validate()

This means that the Form.formData is the model and that we do our validation after ng-model
has updated the model.

davidlgj committed:
Clearing dist from PR

davidlgj commented on source/directives/validate.ts:
```diff
@@ -0,0 +1,41 @@
+        scope.ngModel.$viewChangeListeners.push(function() {
+          console.log('Going to validate ', scope.ngModel.$modelValue)
+
+          return attribute.validate()
```
@bvaughn I updated the validate directive to use the `AttributeMetadata.validate()` instead and also bound all fields error messages against `AttributeMetadata.errorMessages` (although I only show the first.)

bvaughn replied:
Nice!

bvaughn commented:
Going to merge this because I want to play with it some.
